src/Nethermind/Nethermind.JsonRpc/Modules/RpcModuleProvider.cs

```diff
@@ -52,22 +65,36 @@ public RpcModuleProvider(IFileSystem fileSystem, IJsonRpcConfig jsonRpcConfig, I
 
         public IReadOnlyCollection<string> All => _modules;
 
+        private void RegisterNonGeneric(Type moduleType, IRpcModulePool pool)
+        {
+            // Hey its either this of changing like, 5 class.
+            MethodInfo generic = _registerMethod.MakeGenericMethod(moduleType);
+            generic.Invoke(this, [pool]);
+        }
+
         public void Register<T>(IRpcModulePool<T> pool) where T : IRpcModule
         {
-            RpcModuleAttribute attribute = typeof(T).GetCustomAttribute<RpcModuleAttribute>();
+            Type moduleClass = typeof(T);
+            RpcModuleAttribute attribute = moduleClass.GetCustomAttribute<RpcModuleAttribute>();
             if (attribute is null)
             {
-                if (_logger.IsWarn) _logger.Warn($"Cannot register {typeof(T).Name} as a JSON RPC module because it does not have a {nameof(RpcModuleAttribute)} applied.");
+                if (_logger.IsWarn) _logger.Warn($"Cannot register {moduleClass.Name} as a JSON RPC module because it does not have a {nameof(RpcModuleAttribute)} applied.");
                 return;
             }
 
+            if (_logger.IsTrace) _logger.Trace($"Registering module {moduleClass.Name} as part of {attribute.ModuleType} module");
             string moduleType = attribute.ModuleType;
             lock (_updateRegistrationsLock)
             {
+                using var methods = GetMethods<T>(moduleType).ToPooledList(0);
+                var poolRecord = GetPool(pool);
+
                 // FrozenDictionary can't be directly updated (which makes it fast for reading) so we combine the two sets of
                 // data as an Enumerable create an new FrozenDictionary from it and then update the reference
-                _pools = GetPools<T>(moduleType, pool).ToFrozenDictionary(StringComparer.Ordinal);
-                _methods = _methods.Concat(GetMethods<T>(moduleType)).ToFrozenDictionary(StringComparer.Ordinal);
+                _pools = _pools.Concat(methods
+                    .Select(method => new KeyValuePair<string, Pool>(method.Key, poolRecord)))
+                    .ToFrozenDictionary<string, Pool>(StringComparer.Ordinal);
+                _methods = _methods.Concat(methods).ToFrozenDictionary(StringComparer.Ordinal);
```

Comment:
Maybe the methods and pools should be concatenated in normal dictionary and Frozen when first read?

Reply:
Fixed